elasticsearch/src/main/scala/akka/stream/alpakka/elasticsearch/ElasticsearchSinkStage.scala

```diff
@@ -0,0 +1,124 @@
+      override protected def convert(value: JsObject): JsObject = value
+    }
+    (logic, promise.future)
+  }
+
+}
+
+final class ElasticsearchSinkStageTyped[T](indexName: String,
+                                           typeName: String,
+                                           client: RestClient,
+                                           settings: ElasticsearchSinkSettings)(implicit writer: JsonWriter[T])
+    extends GraphStageWithMaterializedValue[SinkShape[IncomingMessage[T]], Future[Done]] {
+
+  val in = Inlet[IncomingMessage[T]]("ElasticsearchSink.in")
+
+  override def shape: SinkShape[IncomingMessage[T]] = SinkShape.of(in)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[Done]) = {
+    val promise = Promise[Done]()
+    val logic = new ElasticsearchSinkLogic(indexName, typeName, client, settings, shape, in, promise) {
+      override protected def convert(value: T): JsObject = value.toJson.asJsObject
+    }
+    (logic, promise.future)
+  }
+
+}
+
+sealed abstract class ElasticsearchSinkLogic[T](indexName: String,
+                                                typeName: String,
+                                                client: RestClient,
+                                                settings: ElasticsearchSinkSettings,
+                                                shape: SinkShape[IncomingMessage[T]],
+                                                in: Inlet[IncomingMessage[T]],
+                                                promise: Promise[Done])
+    extends GraphStageLogic(shape) {
+
+  private val buffer = new util.ArrayDeque[IncomingMessage[T]]()
+
+  protected def convert(value: T): JsObject
+
+  override def preStart(): Unit =
+    pull(in)
+
+  setHandler(in,
+    new InHandler {
+    override def onPush(): Unit = {
+      val message = grab(in)
+      buffer.addLast(message)
+      if (buffer.size >= settings.bufferSize) {
+        sendBulkRequest(buffer.toArray(new Array[IncomingMessage[T]](buffer.size)))
+        buffer.clear()
+      }
+      pull(in)
+    }
+
+    override def onUpstreamFailure(exception: Throwable): Unit = {
+      failStage(exception)
+      promise.tryFailure(exception)
+    }
+
+    override def onUpstreamFinish(): Unit = {
+      if (!buffer.isEmpty) {
+        sendBulkRequest(buffer.toArray(new Array[IncomingMessage[T]](buffer.size)))
+        buffer.clear()
+      }
+      completeStage()
+      promise.trySuccess(Done)
+    }
+
+    private def sendBulkRequest(messages: Seq[IncomingMessage[T]]): Unit =
+      try {
+        val json = messages.map { message =>
+          s"""{"index": {"_index": "${indexName}", "_type": "${typeName}"${message.id.map { id =>
+               s""", "_id": "${id}""""
+             }.getOrElse("")}}
+               |${convert(message.source).toString}""".stripMargin
+        }.mkString("", "\n", "\n")
+
+        client.performRequest(
```

Comment:
Exactly. It is almost Apache HTTP Components, but it offers handling multiple nodes when accessing to Elasticsearch cluster via HTTP.
https://www.elastic.co/guide/en/elasticsearch/client/java-rest/current/_initialization.html

However it seems to be possible to use [CloseableHttpAsyncClient](http://hc.apache.org/httpcomponents-asyncclient-dev/httpasyncclient/apidocs/org/apache/http/impl/nio/client/CloseableHttpAsyncClient.html). Should I use it instead?